### PR TITLE
add support for opening file from memory

### PR DESCRIPTION
**Pull request by thehesiod**
fixes: https://github.com/Unidata/netcdf4-python/issues/406 and https://github.com/Unidata/netcdf4-python/issues/295, requires: https://github.com/Unidata/netcdf-c/issues/394 to be fixed

### CODE REVIEW

**Comment by thehesiod** [comment]
btw unittests are failing because it can't find the nc_open_mem routing, not sure if this should be dynamically looked up or bump up the minimum netCDF version required

**Comment by jswhit** [comment]
This is great, thanks!

To fix the issue with the tests, take a look at the `check_api` function in `setup.py`.  You can add a check there to see if the include file has `nc_open_mem`.  setup.py should then modify `include/constants.pyx` and add `DEF HAS_NC_OPEN_MEM = 1`.  The`nc_open_mem` calls in `_netCDF4.pyx` can then be wrapped with `IF HAS_NC_OPEN_MEM:`.

**Comment by thehesiod** [comment]
btw, why is netCDF4/_netCDF4.c checked in?  it's generated when setup.py is run.

**Comment by jswhit** [comment]
We'll also need a test.

_netCD4.c has to be manually generated.  It should be created with all the values in include/constants.pyx set to 0.  It's only used if cython is not installed.

**Comment by jswhit** [comment]
Probably we should just require cython and get rid of _netCDF.c in the repo.

**Comment by matthew-brett** [comment]
Cython is super-easy to install these days - there are wheels for many versions for OSX, Windows, Linux.

**Comment by thehesiod** [comment]
working on updates, btw mind if I PEP'ify some of the files?  get a lot of warnings in PyCharm

**Comment by thehesiod** [comment]
ok, made nc_open_mem lookup dynamic and added a test.

**Comment by thehesiod** [comment]
btw, I'm not even sure how the .c file here helps because isn't it specific to a python version / platform / cython version?  Not sure how that ever worked correctly.

**Comment by thehesiod** [comment]
btw let me know how I should fix this new unittest error.  I'm guessing it's compiling with all the settings off.  Further it won't work until there's a version of netcdf-c with the fix

**Comment by shoyer** [comment]
> Probably we should just require cython and get rid of _netCDF.c in the repo.

The standard way to do this is to generate C files for upload to pypi, but not include them in the source.
cytoolz has a relatively sane version of this setup: https://github.com/pytoolz/cytoolz/blob/master/setup.py

File: netCDF4/_netCDF4.pyx [hunk 1]
```diff
@@ -1712,7 +1721,7 @@ references to the parent Dataset or Group.
     the parent Dataset or Group.""" 
 
     def __init__(self, filename, mode='r', clobber=True, format='NETCDF4',
```
Comment (shoyer) [inline netCDF4/_netCDF4.pyx]: Can we make `filename` optional if `memory` is provided?

Even if netCDF-C needs a bogus `filename` argument, I imagine we could generate something at random.

Comment (thehesiod) [inline netCDF4/_netCDF4.pyx]: IMHO since `nc_open_mem` takes a filename we should use it

Comment (shoyer) [inline netCDF4/_netCDF4.pyx]: OK, I guess so. It just seems pretty pointless to me :)

Comment (thehesiod) [inline netCDF4/_netCDF4.pyx]: ya, weirdness :)

**Comment by jswhit** [comment]
Looks like the `_netCDF.c` still uses` nc_open_mem`, so that will cause a failing test.  To generate one that will work with the minimum version of the netcdf library, edit `constants.pyx` and set all the elements to zero, then `cd _netCDF; cython -I../include _netCDF4.pyx`, then commit the new version of `_netCDF.c`.

I'd like to remove _netCDF.c from the repo, but that should be a separate pull request.

Are the other failing tests due to the bug in Unidata/netcdf-c#394?

**Comment by thehesiod** [comment]
so I think the tests are failing now because it's running the mem test with the feature disabled?

**Comment by jswhit** [comment]
The test fails on my machine even though I have 4.4.1 (with nc_open_mem).   I guess this is because of Unidata/netcdf-c#394?  I guess we should disable the test in run_all.py if ``__netcdf4libversion__`` < 4.4.1.2 (assuming that the fix gets in the next release of netcdf-c).

**Comment by thehesiod** [comment]
btw, it seems like we should be able to use a memoryview insead reading http://docs.cython.org/en/latest/src/userguide/memoryviews.html?highlight=memory%20views.html.  Let me know what you guys think

File: netCDF4/_netCDF4.pyx [hunk 2]
```diff
@@ -1776,17 +1784,27 @@ references to the parent Dataset or Group.
...
+        
+        **`memory`**: if not `None`, open file with contents taken from this block of memory.
+        Must be a sequence of bytes.  Note this only works with "r" mode.
         """
         cdef int grpid, ierr, numgrps, numdims, numvars
         cdef char *path
         cdef char namstring[NC_MAX_NAME+1]
+
+        memset(&self._buffer, 0, sizeof(self._buffer))
+
```
Comment (jswhit) [inline netCDF4/_netCDF4.pyx]: Probably ought to add `_buffer` to `_private_atts` so it doesn't get mistaken for a netcdf attribute.

**Comment by jswhit** [comment]
Don't see any real benefit to using memoryviews, since we're not slicing or manipulating the buffer in any way.

File: netCDF4/_netCDF4.pyx [hunk 3]
```diff
@@ -1054,9 +1055,11 @@ default_fillvals = {#'S1':NC_FILL_CHAR,
 is_native_little = numpy.dtype('<f4').byteorder == '='
 is_native_big = numpy.dtype('>f4').byteorder == '='
 
-# hard code this here, instead of importing from netcdf.h
+# hard code these here, instead of importing from netcdf.h
 # so it will compile with versions <= 4.2.
 NC_DISKLESS = 0x0008
+NC_INMEMORY = 0x8000
```
Comment (jswhit) [inline netCDF4/_netCDF4.pyx]: Doesn't look like `NC_INMEMORY` is ever used

File: netCDF4/_netCDF4.pyx (hunk 1 above)
Comment (jswhit) [inline netCDF4/_netCDF4.pyx]: Some text should be added to the docstring noting that `filename` is only used to set the value of `filepath` when `memory` is specified.

File: test/tst_open_mem.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import unittest
+import netCDF4
+
+CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
+
+
+class TestOpenMem(unittest.TestCase):
+    def test_mem_open(self):
+        # Needs: https://github.com/Unidata/netcdf-c/pull/400
+        if netCDF4.__netcdf4libversion__ >= '4.4.1.2':
```
Comment (jswhit) [inline test/tst_open_mem.py]: Why not just add this check in run_all.py so the test is skipped altogether?

File: netCDF4/_netCDF4.pyx (hunk 3 above)
Comment (thehesiod) [inline netCDF4/_netCDF4.pyx]: removed

File: test/tst_open_mem.py (hunk 4 above)
Comment (thehesiod) [inline test/tst_open_mem.py]: actually updated the test to test different scenarios of this

File: netCDF4/_netCDF4.pyx (hunk 2 above)
Comment (thehesiod) [inline netCDF4/_netCDF4.pyx]: added

**Comment by jswhit** [comment]
Merging now - thanks again for this contribution!

**Comment by thehesiod** [comment]
thank you for the quick responses!  pleasure!

**Comment by TWellman** [comment]
Is there a working example/documentation based on this discussion for pulling a dataset via a url into memory using the newest netCDF4 version? I've read material on adding a memory=bytes construct in the __init__ function to use the in-memory read capability from nc_open_mem(). However, my initial attempts have produced file location errors - "No such file or directory". I also ran a portion of the test script provided at https://github.com/Unidata/netcdf4-python/blob/master/test/tst_open_mem.py. An error was raised for netCDF4.__netcdf4libversion__ = '4.4.1.1' (< 4.4.1.2). Do I need to pull from https://github.com/Unidata/netcdf-c/pull/400 and integrate netcdf library changes directly into the package? Clarification is appreciated!

**Comment by TWellman** [comment]
I am testing on a Jupyter notebook using Python 3.5 and netCDF4 version 1.3.0, but could switch if needed.

**Comment by jswhit** [comment]
I think you need to link against the updated netcdf-c library.  This one should work:

https://github.com/Unidata/netcdf-c/releases/tag/v4.5.0-rc1

**Comment by dopplershift** [comment]
Another option is to pass it a filename to an empty file, like from `tempfile.NamedTemporaryFile`

**Comment by TWellman** [comment]
Great, substituting a completely unrelated local netCDF file as the "filename" but also passing a memory file in byte format proved to be successful. The returned file dimensions were consistent with the memory file rather than the unrelated local netCDF file with different dimensions/variables. So that checks out.

I also attempted **unsuccessfully** to use the temp file option using FILE_NAME = tempfile.NamedTemporaryFile(suffix='.nc', delete=False).name, mainly because it is much cleaner in my view. However, using a temp file as prescribed returned an OSError: NetCDF: Unknown file format. Should I format the temp file in a specific way? Thank you.

**Comment by thehesiod** [comment]
I think "filename" just needs to be non-empty string

**Comment by TWellman** [comment]
If I just input a text string the code returns an OSError: No such file or directory, e.g. using nc_file = netCDF4.Dataset('temp_file', mode = 'r', memory=bytefile)

**Comment by thehesiod** [comment]
what version of netcdf-c ?

**Comment by dopplershift** [comment]
@TWellman I needed it to be a netCDF file (apparently--I guess I forgot about it). This is what I've done for now until netCDF 4.5 is released:
```python
    with NamedTemporaryFile(suffix='.nc') as temp:
        # Create a temporary netCDF file to work around bug in netCDF C 4.4.1.1
        # We shouldn't actually need any file on disk.
        nc_temp = Dataset(temp.name, 'w').close()
        nc =  Dataset(temp.name, memory=mem_buffer)
```

**Comment by TWellman** [comment]
I should probably upgrade as suggested earlier. I am using netCDF4.netcdf4libversion = '4.4.1.1', netCDF4 version 1.3.0.

**Comment by TWellman** [comment]
Great, thank you!

**Pull request merged**